File: #0403.frog-jump.cpp
class Solution {
public:
    bool canCross(vector<int>& stones) {

        int size = stones.size();
        if (size == 0) {
            return true;
        }
        set<int> s(stones.begin(), stones.end());
        stones = vector<int>(s.begin(), s.end());
        
        int goal = stones.back();
        

        
        vector<vector<bool>> dp(size, vector<bool>(size + 1, false));
        dp[0][1] = true;
        for (int i = 1; i < size; ++i) {
            for (int j = 0; j < i; ++j) {
                int jump = stones[i] - stones[j];
                if (jump >= size + 1 || !dp[j][jump]) {
                    continue;
                }
                
                if (i == size - 1) {
                    return true;
                }
                
                for (int k = jump - 1; k <= jump + 1; k++) {
                    if (k <= 0) {
                        continue;
                    }

                    dp[i][k] = true;
                }
            }
        }
        
        return false;
    }
};

File: #0403.frog-jump.js
/**
 * @param {number[]} stones
 * @return {boolean}
 */
var canCross = function(stones) {
    let q = [[0,0]];
    let length = stones.length - 1;
    let isBool = {};
    while (q.length) {
        let [i,k] = q.shift();
        if(i == length) {
            return true;
        }

        if (k == 0 && i > 0 || isBool[`${i}:${k}`]) {
            continue;
        }

        isBool[`${i}:${k}`] = true;
        let j = i + 1;
        while (stones[j] <= stones[i] + k + 1) {
            if(stones[i] + k == stones[j]) {
                q.push([j,k]);
            }

            if (stones[i] + k + 1 == stones[j]) {
                q.push([j,k + 1]);
            }

            if (stones[i] + k - 1 == stones[j]) {
                q.push([j,k - 1]);
            }

            j++;
        }  
    }   
    return false;
};
